### ISSUE TITLE
Teach strbuf a new trick: STRBUF_INIT_CONST("default")

### ISSUE THREAD

**Issue by dscho** [gitgitgadget/git#398]
From https://public-inbox.org/git/nycvar.QRO.7.76.6.1806210857520.11870@tvgsbejvaqbjf.bet/:

This new way to initialize an `strbuf` would set `buf` to the specified constant string, the appropriate `len = strlen(string)`, and `alloc` to `0`.

That way, we can initialize strbufs with constant values, only allocating memory when/if needed.

We will need to be careful to revisit all `strbuf_*()` functions to make sure that they test not only for enough space when growing the string, but also when reducing it (unless reducing to `len == 0`, in which case we can reassign `strbuf_slopbuf` to the `buf` field).

We also will need to revisit `strbuf_*()` to understand `alloc < len` to mean that we cannot call `realloc()`, but have to `malloc() && memcpy()` instead.

**Comment by mo7sener** [comment]
Can I take that?

**Comment by dscho** [comment]
> Can I take that?

@mo7sener of course! No need to ask for permission, this is open source, baby :grinning:

**Comment by mo7sener** [comment]
Ok, cool. I got a question, though. When you say:
> We will need to be careful to revisit all `strbuf_*()` functions to make sure that they test not only for enough space when growing the string, but also when reducing it (unless reducing to `len == 0`, in which case we can reassign `strbuf_slopbuf` to the `buf` field).
I don't quite get why we need to test why the `strbuf` has enough space when shrinking. Excuse me if that's a stupid question, as I'm still a newbie 😄 .

**Comment by dscho** [comment]
@mo7sener I _think_ what I meant four months ago was that constant strings are considered immutable. So you cannot just shorten it by setting a byte in the middle to `NUL`. Which means that you need to copy, and for that, you need to allocate, even if you are shrinking the string, _iff_ it was initialized via `STRBUF_INIT_CONST()`.

**Comment by mo7sener** [comment]
@dscho So, just to be sure, does that seem about right?
```C
void strbuf_make_var(struct strbuf *sb)
{
    char* str_cpy;
    ALLOC_ARRAY(str_cpy, sb->len);
    memcpy(str_cpy, sb->buf, sb->len);
    strbuf_attach(sb, str_cpy, sb->len, sb->len+1);
}
```
This is the function that would turn a const `strbuf` to a variable one.

**Comment by dscho** [comment]
> ```c
>     ALLOC_ARRAY(str_cpy, sb->len);
> ```

I think this should have `sb->len + 1`, otherwise it looks correct. This function should be called by `strbuf_grow()` and by `strbuf_setlen()`.

**Comment by mo7sener** [comment]
@dscho it should be called by other functions that change length, right? Like `strbuf_trim`?

**Comment by dscho** [comment]
> it should be called by other functions that change length, right?

Yes. Or for that matter, any function that wants to modify the bytes in the string.

**Comment by adlternative** [comment]
@dscho  This should be an issue from ancient times,haha!I have some ideas , but I don't know if I was right ,Can you help me see if there is a problem with my thinking?#824
There may be some bugs... :)
